src/util/boolset.rs

```diff
@@ -0,0 +1,78 @@
+use std::cmp::Eq;
+use std::collections::HashMap;
+use std::default::Default;
+use std::fmt::Debug;
+use std::hash::Hash;
+
+// Like a hash table, the bool set is a data structure where keys
+// can be explicitly set to true or false.
+
+#[derive(Debug, Default, Clone, PartialEq)]
+pub struct BoolSet<T>
+where
+    T: Hash + Eq + Default + Clone,
+{
+    map: HashMap<T, bool>,
+    default_val: bool,
+}
+
+impl<T> BoolSet<T>
+where
+    T: Hash + Eq + Default + Clone,
+{
+    pub fn new_with_default(default_val: bool) -> Self {
+        BoolSet {
+            map: HashMap::new(),
+            default_val,
+        }
+    }
+
+    // Maps the key to the value in bool set, overwriting any existing mapping for the key.
+    pub fn insert(&mut self, key: T, val: bool) {
+        self.map.insert(key, val);
+    }
+
+    // Returns the value for the key in bool set.
+    // If no value is found for the key, then default value is returned.
+    pub fn get(&self, key: &T) -> bool {
+        if let Some(val) = self.map.get(key) {
+            val.to_owned()
+        } else {
+            self.default_val
+        }
+    }
+
+    pub fn update(&mut self, other: &Self) {
+        self.map.extend(other.map.clone());
+        self.default_val = other.default_val;
```

Comment:
This is a bit sketchy, should we just assert that the default values match?

Do we really need this rather than just a map and using `.unwrap_or(true)` / `.unwrap_or(false)`?

Reply:
To avoid writing too many duplicated `.unwrap_or(true)` / `.unwrap_or(false)`, bool set might be useful.

If we need to assert that the default values match, should we do that at compile time or runtime?

Reply:
It seems to do that at compile time you'd need to somehow make it generic, or just fix the default value to true, which is the only thing that we're using right now I think? The first is probably not worth it.

Otherwise, just `assert_eq!(self.default_val, other.default_val, "Can't merge BoolSet with different default values")` or such.